### PR TITLE
[DPHyp] Draft added

### PR DESCRIPTION
**Pull request by pashandor789**
### Changelog entry 

...

### Changelog category 

* New feature
* Experimental feature
* Improvement
* Performance improvement
* Bugfix 
* Backward incompatible change
* Documentation (changelog entry is not required)
* Not for changelog (changelog entry is not required)

### Additional information

...

### CODE REVIEW

File: ydb/library/yql/dq/opt/dphyp_bitset.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    return (lhs & rhs) != 0;
+}
+
+template <typename TNodeSet>
+inline bool IsSubset(const TNodeSet& lhs, const TNodeSet& rhs) {
+    return (lhs & rhs) == lhs;
+}
+
+template <typename TNodeSet>
+inline bool HasSingleBit(TNodeSet nodeSet) {
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dphyp_bitset.h:20]: HasSingleBit можно ж != 0 просто проверить?

File: ydb/library/yql/dq/opt/bitset.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        if (nodeSet[i]) {
+            return i;
+        }
+    }
+
+    Y_ASSERT(false);
+    return nodeSet.size();
+}
+
+template <typename TNodeSet>
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/bitset.h:34]: Need a comment for this iterator, is it doing anything else than just iterating? And where is it used?

File: ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
+#include "dq_opt_conflict_rules_collector.h"
+#include <util/generic/hash_set.h>
+
+namespace NYql::NDq {
+
+/* To make ASSOC, RASSCOM, LASSCOM tables simplier */
+EJoinKind GetEquivalentByAlgebraicPropertiesJoin(EJoinKind joinKind) {
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp:7]: The name is broken English

Better name: GetEquivalentJoinByAlgebraicProperties

File: ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        case EJoinKind::Exclusion:
+            return EJoinKind::InnerJoin;
+        case EJoinKind::LeftOnly:
+            return EJoinKind::LeftJoin;
+        default:
+            return joinKind;
+    }    
+}
+
+bool OperatorIsCommut(EJoinKind joinKind) {
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp:18]: OperatorIsCommutative

You saved a couple of chars for no good reason

File: ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        case EJoinKind::Cross:
+            return true;
+        default:
+            return false;
+    }
+
+    Y_UNREACHABLE();
+}
+
+bool OperatorsAreAssoc(EJoinKind lhs, EJoinKind rhs) {
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp:32]: OperatorsAreAssociative

File: ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    };
+
+    if (!(ASSOC_TABLE.contains(lhs))) {
+        return false;
+    }
+
+    return ASSOC_TABLE[lhs].contains(rhs);
+}
+
+bool OperatorsAreLeftAsscom(EJoinKind lhs, EJoinKind rhs) {
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp:50]: Write a short header specifying what LeftAsscom means, or just put in a reference to the paper

File: ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    };
+
+    if (!(LASSCOM_TABLE.contains(lhs))) {
+        return false;
+    }
+
+    return LASSCOM_TABLE[lhs].contains(rhs);
+}
+
+bool OperatorsAreRightAsscom(EJoinKind lhs, EJoinKind rhs) {
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.cpp:69]: Same here

File: ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.h [hunk 8]
```diff
@@ -0,0 +1,141 @@
+#pragma once
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_conflict_rules_collector.h:1]: You need to have a header comment here specifying what this conflict resultion is all about.
A single paragraph with a link to the paper will do

File: ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+ * DPHyp (Dynamic Programming with Hypergraph) is a graph-aware
+ * join eumeration algorithm that only considers CSGs (Connected Sub-Graphs) of
+ * the join graph and computes CMPs (Complement pairs) that are also connected
+ * subgraphs of the join graph. It enumerates CSGs in the order, such that subsets
+ * are enumerated first and no duplicates are ever enumerated. Then, for each emitted
+ * CSG it computes the complements with the same conditions - they much already be
+ * present in the dynamic programming table and no pair should be enumerated twice.
+ * Details are described in white papper - "Dynamic Programming Strikes Back".
+ */
+template <typename TNodeSet>
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h:33]: Add a comment why this class is templated by TNodeSet

File: ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+        } else if (res[i] == 0 && carry) {
+            res[i] = 1;
+            carry = false;
+        }
+    }
+
+    return res;
+}
+
+template<typename TNodeSet> std::shared_ptr<TJoinOptimizerNodeInternal> TDPHypSolver<TNodeSet>::Solve() {
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h:216]: Need a comment for this method. The name 'solve' is from a paper and is a poor description of what happens here

File: ydb/library/yql/dq/opt/dq_opt_make_join_hypergraph.h [hunk 11]
```diff
@@ -0,0 +1,81 @@
+#pragma once
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_make_join_hypergraph.h:1]: Need a header comment about contruction of hypergraphs.

**Review by pavelvelikhov — CHANGES_REQUESTED** [review]

File: ydb/library/yql/dq/opt/bitset.h [hunk 12]
```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <stdlib.h>
+
+/* 
+ * This header contains helper functions for working with bitsets.
+ * They are templated by TNodeSet, which is a std::bitset<>.
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/bitset.h:7]: + We use the the template for efficiency: for 64 bit nodesets we implement a faster next subset functionality

File: ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
```
Comment (pavelvelikhov) [inline ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h:220]: I didn't find the comment for Solve method...

**Review by pavelvelikhov — CHANGES_REQUESTED** [review]

File: ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h (hunk 13 above)
Comment (pashandor789) [reply ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h:220]: it it written in class definition

**Pull request closed without merging**